### ISSUE TITLE
`include_initial` changefeeds appear to sometimes skip values inserted at the batch boundary

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#5216]
This is easiest to run into with very small batch sizes and lots of duplicate sindex values.  The file `init_trivial_fuzz.rb` that will be merged into `next` and `2.2.x` shortly can also be used to reproduce this if you compare the `$state` document at the end to the actual state of documents in the table.

I suspect this is caused by `truncate_and_get_left` being too liberal in its calculation of the currently-read range.

@danielmewes -- I'll try to get this done over the holidays, but as far as I can tell this isn't a new bug so it shouldn't hold up the release of 2.2.

**Comment by danielmewes** [comment]
I can reproduce this even with a simple `tbl.changes()` in the `init_trivial_fuzz.rb` test.

Luckily it looks like the issue disappears if there is at least one document in the table for each hash shard when initiating the query. It seems like if a hash shard is initially empty, we miss any changes that happen to that hash shard during the initial data stream.

I'm going to investigate this a bit more until @mlucy can pick things up again.

**Comment by danielmewes** [comment]
I'd like to try getting this into 2.2.2, so we don't have to do another point release immediately after.

**Comment by danielmewes** [comment]
The problem appears to be that our `active_range_t` is missing the empty hash shards.
`splice_stream_t` seems to assume that `active_range_t` always covers the full domain though.

`splice_stream_t::discard()` even has an `r_sanity_fail()` that should trigger if this isn't the case. However in practice this often doesn't work, because the change that's outside the initial hash shards will be filtered out earlier in that function if it's past the `right_fencepost`.

I'll make an attempt at fixing this. We'll need to wait for @mlucy to get a chance at reviewing the change before we can merge it though.

**Comment by danielmewes** [comment]
Hmm my understanding of this was a bit off, since I was conflating different range structures. The `ranges` in `discard` are not completely equivalent to the active ranges, and they don't even consider hash shards. So above explanation is probably a bit off.

**Comment by mlucy** [comment]
I took a look at this and I think there are two problems: the first is that the `active_ranges_t` has no entries for the empty hash shards, but the other one (which is harder to catch with the test script) is that reads issued after truncating sometimes have ranges that are too small.

Consider the case where you're ordering on field `A` and have two shards.  Your first read returns:

Shard 1: [1, 3, 5, 7] (remaining range: (7, INF))
Shard 2: [2, 4] (remaining range: (4, INF))

We pop off `[1, 2, 3, 4]`, but we can't pop off `5` because there might be a `4.5` on shard 2 that we didn't read.  So we're left with:

Shard 1: [5, 7] (remaining range: (7, INF))
Shard 2: [] (remaining range: (4, INF))

Then we do the truncate step, and we get:

Shard 1: [] (remaining range: (5, INF))
Shard 2: [] (remaining range: (4, INF))

While returning a `last_read_range` of (-INF, 4].

The problem is that when we issue the next read, it's still issued on a per-shard basis, so if between the two reads a value is inserted at `4.5` on Shard 1, we'll miss it.

We could patch this to work by updating the remaining range for Shard 1 to `(4, INF)` (and making sure it does something correct if you have multiple range shards), but I think that the underlying problem here is that we introduced a needlessly complicated range-recalculation step because we didn't want to have to make the changefeed logic shard-aware for the release.  I think that was a mistake in retrospect: we should probably fix this by making the changefeed logic track active ranges on a per-shard basis.

**Comment by danielmewes** [comment]
Yeah I noticed the same second issue during testing.

I just got a version that's at least passing all the tests, but there are a few things going on that I don't fully understand yet. Specifically I'm sometimes getting weird active ranges that lie outside the shard boundaries, and I'm not quite sure why that's happening. I can clip the ranges, and that appears to work, but there are obviously some things going on that I don't understand which makes me fairly uncomfortable.

I'll look into this more tomorrow and will see if I can figure out those things.
I'd *really* like to ship this with 2.2.2, since include initial changefeeds seem pretty seriously broken right now. However I'm not happy with my patches yet.

@mlucy You can find my patch in branch daniel_5216, in commit 8bb8936e75b250e56551758cbc0f56859fd0b883 if you'd like to take a look.

**Comment by mlucy** [comment]
@danielmewes -- I didn't get much time to look at the patch, but I added a comment on one thing that seemed obviously wrong.  In general the patch looked complicated and likely to be subtly incorrect (which, to be fair, is better than the current implementation which is blatantly incorrect).

I really think it would be better to patch this by making the changefeed logic aware of shards.  Here's roughly what that would look like:
* Add a way to correlate `region_t`s with the changefeed `uuid_u`s.
  - This is actually really easy, because the read performs a changefeed stamp at the same time, so we just need to combine the `<uuid_u, stamp>` map and the `<region_t, DATA>` map in the read response into a `<region_t, (uuid_u, stamp, DATA)>` map (since there's a one-to-one mapping between `region_t`s and changefeed server UUIDs).  We'd probably have to maintain this relationship during unsharding, but that should be easy.
  - Once we have this, we'll store the `uuid_u` and `stamp` values on the `hash_range_with_cache_t` type instead of in the `shard_stamps` map.
* Replace `truncate_and_get_active_state` with `get_active_state`, which doesn't truncate and returns one range per hash shard.
  - The `active_state_t` type would change from returning a key range and a `<uuid_u, stamp>` map to returning a `<uuid_u, (stamp, range)>` map with one entry per hash shard, where the ranges can be different for different hash shards.  You could get this by iterating over the `active_ranges`.
  - To make this work correctly you would have to make sure we're doing an UNORDERED traversal so that all the data is always read out of the map, which will make it safe to not truncate.  (As a result you should be able to assert that all ranges are either ACTIVE or EXHAUSTED.)
* We're already using an `std::map<uuid_u, stamped_range_t>` to do the splicing in `splice_stream_t`, so that should require minimal changes: we just update each of the entries in that map separately based on the return value of `get_active_state`.

It would probably turn out to be about the same amount of code, and it would be much more obviously correct.

**Comment by mlucy** [comment]
(That said, if you already have a patch that's strictly better than the current state, it wouldn't hurt to release that in 2.2 and then do the more obviously-correct fix for another point release if it ends up being hard for some reason.)

**Comment by mlucy** [comment]
(Also, you probably already thought of this, but it would probably be valuable to run the tests that reproduce this bug with 4 range shards and on both primary and secondary indexes.)

**Comment by danielmewes** [comment]
Thanks for the great description of the shard-based solution. I'll probably give that a shot tomorrow and see how far I can get. It indeed doesn't sound too complicated, and would probably work better.

**Comment by danielmewes** [comment]
A fix is up in review 3404 with @mlucy (branch: `daniel_5216_sharding`).
It fixes the issue "the right way" by making include initial changefeeds splice each shard individually.

It also fixes another bug that made changefeed stamps sometimes be out of order.

**Comment by danielmewes** [comment]
Fixed in v2.2.x f4ffd12 and in next bf212e5.
